src/modules/wasi/x86-64-linux/WspOneModule-x86-64-linux.v3

```diff
@@ -88,38 +95,104 @@ class X86_64Linux_WspOneModule extends WspOneModule {
 	def host_random_get(region: Range<byte>) -> HostResult {
 		var t = Linux.syscall(SYS_getrandom, (Pointer.atContents(region), region.length, 0));
 		if (t.0 < 0) return i(mapErrno(t.0));
-		return HostResult.Value1(Values.I32_0);
+		return SUCCESS;
 	}
 	def clock_res_get(args: Range<Value>) -> HostResult {
 		var clock_id = mapClockId(Values.unbox_u(args[0]));
-		if (clock_id < 0) return HostResult.Value1(Values.box_i(WspOneErrno.INVAL.code));
+		if (clock_id < 0) return i(WspOneErrno.INVAL.code);
 		var ptr = Values.unbox_u(args[1]);
 		var range = memory.range_ol_32(ptr, 8);
 		if (range.reason != TrapReason.NONE) return HostResult.Throw(Trap.new(range.reason, null, null));
 		var t = Linux.syscall(SYS_clock_get_res, (clock_id, Pointer.atContents(timespec)));
 		if (t.0 < 0) return i(mapErrno(t.0));
 		DataWriters.write_range_i64(range.result, timespec[0] * 1000000000L + timespec[1]);
-		return HostResult.Value1(Values.I32_0);
+		return SUCCESS;
 	}
 	def clock_time_get(args: Range<Value>) -> HostResult {
 		var clock_id = mapClockId(Values.unbox_u(args[0]));
-		if (clock_id < 0) return HostResult.Value1(Values.box_i(WspOneErrno.INVAL.code));
+		if (clock_id < 0) return i(WspOneErrno.INVAL.code);
 		var lag = Values.unbox_l(args[1]);
 		var ptr = Values.unbox_u(args[2]);
 		var range = memory.range_ol_32(ptr, 8);
 		if (range.reason != TrapReason.NONE) return HostResult.Throw(Trap.new(range.reason, null, null));
 		var t = Linux.syscall(SYS_clock_get_time, (clock_id, Pointer.atContents(timespec)));
 		if (t.0 < 0) return i(mapErrno(t.0));
 		DataWriters.write_range_i64(range.result, timespec[0] * 1000000000L + timespec[1]);
-		return HostResult.Value1(Values.I32_0);
+		return SUCCESS;
+	}
+	def fd_read(args: Range<Value>) -> HostResult {
+		var fd = Values.unbox_i(args[0]);
+		var iovs_ptr = Values.unbox_u(args[1]);
+		var iovs_len = Values.unbox_u(args[2]);
+		var size = Values.unbox_u(args[3]);
+
+		var wasi_fd = fdmap.get(fd);
+		if (wasi_fd == null) return i(WspOneErrno.BADF.code);
+
+		var iovRange = memory.range_ol_32(iovs_ptr, iovs_len * 8);
+		if (iovRange.reason != TrapReason.NONE) return HostResult.Throw(Trap.new(iovRange.reason, null, null));
+
+		var sizeRange = memory.range_ol_32(size, 4);
+		if (sizeRange.reason != TrapReason.NONE) return HostResult.Throw(Trap.new(sizeRange.reason, null, null));
+
+		if (iovs_len > IOV_MAX) return i(WspOneErrno.INVAL.code);
+
+		var iov = Array<byte>.new(int.!(iovs_len * 16));
+
+		for (j < int.!(iovs_len)) {
```

Comment:
Is it possible to factor out a common routine for the iovecs copy?

Reply:
I'm not sure how to handle this return:
https://github.com/titzer/wizard-engine/blob/07a1b0b40cce7cd3db899ae7f71fcdfff064f40f/src/modules/wasi/x86-64-linux/WspOneModule-x86-64-linux.v3#L145

Reply:
Ah, I see. I think you could return a `MaybeTrap<Array<byte>>` from the utility function.